### ISSUE TITLE
Complex effects (interactions, polynomials, etc.) in writeJagsModel.xxx()

### ISSUE THREAD

**Issue by j-e-dalton** [nutterb/HydeNet#26]
This might be a toughie.

The interactions don't seem to be getting converted to JAGS code from inputted R model objects.  See code below for illustration.  I think the problem is within writeJagsModel.lm(), writeJagsModel.glm(), and writeJagsModel.multinom(), specifically in the way the model objects are used to parse out terms --line 10 in writeJagsModel.lm() seems to be agnostic to interactions.

A natural extension of this issue is with polynomial terms, interactions between continuous/polynomial terms and factors, etc., etc., etc.  However, why don't we constrain the issue for the beta release to merely making sure that a) interactions and b) polynomial terms are supported, unless a general solution is relatively easily identified.

I will add a little section warning the user on the limited scope of writeJagsModel.xxx, and that setNode can be used in place when complex models need to be specified.

g1 <- lm(wells ~ 1, data=PE)
g2 <- glm(pe ~ wells, data=PE, family="binomial")
g3 <- lm(d.dimer ~ pe + pregnant, data=PE)
g4 <- xtabs(~ pregnant, data=PE)
g5 <- glm(angio ~ pe, data=PE, family="binomial")
g6 <- glm(treat ~ d.dimer + angio, data=PE, family="binomial")
g7 <- glm(death ~ pe + treat, data=PE, family="binomial")

bagOfModels <- list(g1,g2,g3,g4,g5,g6,g7)

bagNet <- HydeNetwork(bagOfModels)
writeNetworkModel(bagNet, pretty=TRUE)

model{
   wells ~ dnorm(wells ~ 3.7941999999999, 0.630504251834359)
   pe ~ dbern(pe ~ ilogit(-3.90355 + 0.5757_wells))
   d.dimer ~ dnorm(d.dimer ~ 210.24251 + 68.37938_(pe==2) + 29.29496_(pregnant==2), 0.0334725036145318)
   pi.pregnant[1] <- 0.9014; pi.pregnant[2] <- 0.0986
pregnant ~ dcat(pi.pregnant)
   angio ~ dbern(angio ~ ilogit(-2.22585 + 3.28411_(pe==2)))
   treat ~ dbern(treat ~ ilogit(-5.89316 + 0.01994_d.dimer + 1.73354_(angio==2)))
   death ~ dbern(death ~ ilogit(-4.18763 + 5.48082_(pe==2) + -1.93576_(treat==2)))
}

new.DDimer.Model <- lm(d.dimer ~ pe \* pregnant, data=PE)
bagNet <- setNodeModels(bagNet, new.DDimer.Model)

writeNetworkModel(bagNet, pretty=TRUE)

model{
   wells ~ dnorm(wells ~ 3.7941999999999, 0.630504251834359)
   pe ~ dbern(pe ~ ilogit(-3.90355 + 0.5757_wells))
   d.dimer ~ dnorm(d.dimer ~ 210.033541553503, 0.0335041033258058)  ##no interaction!
   pi.pregnant[1] <- 0.9014; pi.pregnant[2] <- 0.0986
pregnant ~ dcat(pi.pregnant)
   angio ~ dbern(angio ~ ilogit(-2.22585 + 3.28411_(pe==2)))
   treat ~ dbern(treat ~ ilogit(-5.89316 + 0.01994_d.dimer + 1.73354_(angio==2)))
   death ~ dbern(death ~ ilogit(-4.18763 + 5.48082_(pe==2) + -1.93576_(treat==2)))
}

**Comment by nutterb** [comment]
I think we may have a solution to this in the `broom` package released to CRAN last month.  It will take a bit of rewriting on my part, but Robinson managed to do something I've been struggling to figure out for years.

**Comment by j-e-dalton** [comment]
I’m back.

Looks appealing.  I like the example in their vignette using nonlinear least squares.  If we go this route, we may eventually be interested in implementing this as well.

I thought I’d fit a polynomial/interaction model and see what it looked like:

require(HydeNet)
require(broom)

data(PE)

g <- glm(pe ~ poly(d.dimer,2)*angio, family="binomial", data=PE)
tidy(g)

![image](https://cloud.githubusercontent.com/assets/7571400/6784061/de70ba2c-d153-11e4-944f-5f37ac1c175a.png)

**Comment by nutterb** [comment]
Now, if I understand this correctly, the formula I would want to implement is

`pe ~ -4.12 + 268.92*d.dimer - 0.83 * d.dimer * 3.05 * [angio == 2] + 22.17 * d.dimer * [angio==2] + 2.07 * d.dimer * [angio==2]`

Is that correct?

**Comment by j-e-dalton** [comment]
You didn't have the squares for the two terms involving d.dimer^2

`pe ~ -4.12 + 268.92*d.dimer - 0.83 * (d.dimer^2) + 3.05 * [angio == 2] + 22.17 * d.dimer * [angio==2] + 2.07 * (d.dimer^2) * [angio==2]`

**Comment by nutterb** [comment]
The code below works.

``` r
library(HydeNet)
g1 <- lm(wells ~ 1, data=PE)
g2 <- glm(pe ~ wells, data=PE, family="binomial")
g3 <- lm(d.dimer ~ pe + pregnant, data=PE)
g4 <- xtabs(~ pregnant, data=PE)
g5 <- glm(angio ~ pe, data=PE, family="binomial")
g6 <- glm(treat ~ d.dimer + angio, data=PE, family="binomial")
g7 <- glm(death ~ pe + treat, data=PE, family="binomial")

bagOfModels <- list(g1,g2,g3,g4,g5,g6,g7)

bagNet <- HydeNetwork(bagOfModels)
bagNet <- setNode(bagNet,
                  treat,
                  nodeFormula = treat ~ poly(d.dimer, 2) + angio,
                  nodeFitter='glm',
                  fitterArgs = list(family=binomial),
                  p = fromData(),
                  fitModel=TRUE)
writeNetworkModel(bagNet, pretty=TRUE)
```

I've generated two bugs, however.  When I use `fitModel=FALSE`, `writeNetworkModel` fails.

The second bug is strange.  if I use `print(bagNet)`, it prints just fine.  but if I just type `bagNet`, it crashes the system.  That concerns me a bit.

Also, we'll have to be careful about passing model objects, it seems.  At the moment I don't know that I have a good way to tell that `d.dimer` is the same node as `poly(d.dimer, 2)`.  If you pass the formula in my `setNode` above in the original model object, it creates two nodes for `d.dimer`.  In theory, it wouldn't be hard to strip out functions, I suppose, but how far down that rabbit hole do we want to go?  Do we want to accommodate nested functions?  will we take just any function?  What if it is a function of two nodes?  I'm going to have to think on that one a little bit.

**Comment by nutterb** [comment]
One more note on this: I haven't incorporated polynomial terms in multinomial logistic regressions.  Mostly because there isn't a `tidy` method for these models.  I suspect I'll have to write that method before I can do the polynomial terms in `HydeNet`.

**Comment by j-e-dalton** [comment]
Note also that your code above involves a main-effects model, not one that included the interaction term poly(d.dimer,2) : angio

**Comment by j-e-dalton** [comment]
I suggest we cover the rabbit hole with mulch and move on to decision & utility nodes.  That's what I would do at home anyways.

**Comment by nutterb** [comment]
``` r
library(HydeNet)
g1 <- lm(wells ~ 1, data=PE)
g2 <- glm(pe ~ wells, data=PE, family="binomial")
g3 <- lm(d.dimer ~ pe + pregnant, data=PE)
g4 <- xtabs(~ pregnant, data=PE)
g5 <- glm(angio ~ pe, data=PE, family="binomial")
g6 <- glm(treat ~ d.dimer + angio, data=PE, family="binomial")
g7 <- glm(death ~ pe + treat, data=PE, family="binomial")

bagOfModels <- list(g1,g2,g3,g4,g5,g6,g7)

bagNet <- HydeNetwork(bagOfModels)
bagNet <- setNode(bagNet,
                  treat,
                  nodeFormula = treat ~ poly(d.dimer, 2) * angio,
                  p = fromData())

writeNetworkModel(bagNet, pretty=TRUE)
```

It would have worked with the interaction if I remembered when I was using strings and when I was using regular expressions.

**Comment by jarrod-dalton** [comment]
Reopening this issue, to be addressed after v1.0.  And below is another example, with which I don't know what to do.

```
library(HydeNet)
net <- HydeNetwork(~ card1.ace | card1
                   + card2.ace | card2
                   + initialPoints | card1*card2*card1.ace*card2.ace
                   + hit1 | initialPoints*dealerUpcard
                   + card3 | hit1
                   + card3.ace | card3
                   + pointsAfterCard3 | initialPoints*card3*card3.ace
                   + hit2 | pointsAfterCard3*dealerUpcard
                   + card4 | hit2
                   + card4.ace | card4
                   + pointsAfterCard4 | pointsAfterCard3*card4*card4.ace
                   + hit3 | pointsAfterCard4*dealerUpcard
                   + card5 | hit3
                   + card5.ace | card5
                   + pointsAfterCard5 | pointsAfterCard4*card5*card5.ace
                   + playerFinalPoints | initialPoints*hit1*pointsAfterCard3
                   *hit2*pointsAfterCard4*hit3*pointsAfterCard5
                   + dealerFinalPoints | dealerUpcard
                   + payoff | playerFinalPoints*dealerFinalPoints)

data(BlackJackTrain)

glm.hit1 <- glm(hit1 ~ initialPoints + I(dealerUpcard %in% c("9","10-K","A")),
                data = BlackJackTrain, family="binomial")
glm.hit2 <- glm(hit2 ~ pointsAfterCard3 + I(dealerUpcard %in% c("9","10-K","A")),
                data = BlackJackTrain, family="binomial")
glm.hit3 <- glm(hit3 ~ pointsAfterCard4 + I(dealerUpcard %in% c("9","10-K","A")),
                data = BlackJackTrain, family="binomial")

net <- setNodeModels(net, glm.hit1, glm.hit2, glm.hit3)
```

```
Error in parse(text = x) : <text>:1:77: unexpected numeric constant
1: hit1 ~ ilogit(8.16617   + 1.93073 * I(dealerUpcard %in% c("9", "10-K", "A"))TRUE
```

**Comment by nutterb** [comment]
This is caused be `HydeNet` not recognizing the node `I(dealerUpcard %in% c("9", "10-k", "A"))`.  This was one of my big fears about translating the formulas is that it can be a bit difficult to strip out the node name from within a function passed to a model.  Eventually, I think I can reach a point where if I assume exactly one node name in a function, I can do it effectively.  

But that's what is happening here.
